scripts/west_commands/completion/west-completion.fish

```diff
@@ -196,21 +196,52 @@ function __zephyr_west_complete_help
 end
 
 function __zephyr_west_complete_board
-    # HWMv1
-    set -l boards (west 2>/dev/null boards --format="{name} {arch}")
-    for board in $boards
-        set -l b (string split " " $board)
-        printf "%s\n" $b[1]\t"$b[2]"
+    set -l one_week_in_s 604800
+    set -l is_cache_valid 0
+
+    set -l cache_folder $HOME/.cache/west
```

Comment:
I don't think `$HOME/.cache` is appropriate for Mac users.

Perhaps a folder under `$__fish_user_data_dir` is a better choice for fish users across OS'es.
Or the temp dir, which would be cleared on reboot.

Reply:
note: temp is not cleared on reboot on windows

Reply:
Indeed it's a problem if it doesn't work on other OS'es. But I am not sure that using `$__fish_user_data_dir` is the correct way to do. It's not well documented but it seems to be a directory used to load third-party software fish functions.
Here is what the documentation say:
> Directories for other software to put their own functions. These are in the directories under $__fish_user_data_dir (usually ~/.local/share/fish, controlled by the XDG_DATA_HOME environment variable) and in the XDG_DATA_DIRS environment variable, in a subdirectory called fish/vendor_functions.d.

Instead I propose to go OS specific, for Linux use `~/.cache/west`, for macOS use `~/Library/Caches/west` (it seems to be the right place but I don't have a macOS setup to check myself). For Windows I think it's fine to use the same as Linux. fish doesn't support Windows "natively", it's only supported through Cygwin, MSYS2 or WSL.

Reply:
Indeed it's a problem if it doesn't work on other OS'es. But I am not sure that using `$__fish_user_data_dir` is the correct way to do. It's not well documented but it seems to be a directory used to load third-party software fish functions.
Here is what the documentation say:
> Directories for other software to put their own functions. These are in the directories under $__fish_user_data_dir (usually ~/.local/share/fish, controlled by the XDG_DATA_HOME environment variable) and in the XDG_DATA_DIRS environment variable, in a subdirectory called fish/vendor_functions.d.

Instead I propose to go OS specific, for Linux use `~/.cache/west`, for macOS use `~/Library/Caches/west` (it seems to be the right place but I don't have a macOS setup to check myself). For Windows I think it's fine to use the same as Linux. fish doesn't support Windows "natively", it's only supported through Cygwin, MSYS2 or WSL.

Reply:
> But I am not sure that using $__fish_user_data_dir is the correct way to do.

Ref: https://fishshell.com/docs/current/completions.html#where-to-put-completions
Ref: https://github.com/fish-shell/fish-shell/pull/8887

> By default, Fish searches the following for completions, using the first available file that it finds:
> ...
> A user-specified directory for third-party vendor completions, usually ~/.local/share/fish/vendor_completions.d (controlled by the XDG_DATA_HOME environment variable);


So `$__fish_user_data_dir/vendor_completions.d` is actually the right place, as this is related to completion handling.

And then let fish itself decide how it will define `$__fish_user_data_dir` (`XDG_DATA_HOME` on Linux) across OS'es.

Reply:
Sorry I still doesn't agree on that, as I understand those lines, the folder is for the completions themselves, not some "random" files. But at the same time I did not find any example of fish completion that use files.

As a user I appreciate when software use existing "system" directories and doesn't put files everywhere on the system, especially in the home directory.

If that's fine for you, and as long as someone come with examples or some best practices that I may not be aware of, I would like to keep the updated solution that use per OS cache folder.

Reply:
> the folder is for the completions themselves, not some "random" files.

True, but I kind of see this caching mechanism as part of the completion script.
One thing though, the user may not have added the completion script to that folder in which case it can be argued that we shouldn't place a cache file there.

> I would like to keep the updated solution that use per OS cache folder.

In such case, then `XDG_HOME_DIR` should as minimum be honored on Linux.

Perhaps the best location is under `<west-workspace>/.west/fish/<completion-cache-file>`.
That will place it together with other west related settings for the workspace, and the `<workspace>/.west` folder is always created by `west`.

Ref:
> Local: Settings in this file affect west’s behavior for the current [west workspace](https://docs.zephyrproject.org/latest/glossary.html#term-west-workspace). The file is `.west/config`, relative to the workspace’s root directory.
https://docs.zephyrproject.org/latest/develop/west/config.html#west-configuration-files

Remember, users have the possibility to swap manifest inside a workspace, for example like this:
```
$ west config manifest.path zephyr # Use vanilla Zephyr manifest 
$ west update

$ west config manifest.path nrf # Use nRF Connect SDK manifest 
$ west update
```

Reply:
> In such case, then XDG_HOME_DIR should as minimum be honored on Linux.

You're right, I updated the code

> Perhaps the best location is under <west-workspace>/.west/fish/<completion-cache-file>.
That will place it together with other west related settings for the workspace, and the <workspace>/.west folder is always created by west.

That may be the best but the `.west` folder isn't created by default by west, should fish create it if it doesn't exist?

Reply:
> That may be the best but the .west folder isn't created by default by west, should fish create it if it doesn't exist?

`.west` folder is created by west when doing a `west init`, and remember that both `west boards` and `west build` are extension commands defined in the west workspace.

Remember, the west workspace folder is not a git folder, it's one level above Zephyr in a standard install.

> If you’ve followed the upstream Zephyr getting started guide, your workspace looks like this:
> ```
> zephyrproject/                 # west topdir
>├── .west/                     # marks the location of the topdir
>│   └── config                 # per-workspace local configuration file
>│
>│   # The manifest repository, never modified by west after creation:
>├── zephyr/                    # .git/ repo
>│   ├── west.yml               # manifest file
>│   └── [... other files ...]
>│
>│   # Projects managed by west:
>├── modules/
>│   └── lib/
>│       └── zcbor/             # .git/ project
>├── net-tools/                 # .git/ project
>└── [ ... other projects ...]
> ```
Ref: https://docs.zephyrproject.org/latest/develop/west/basics.html#example-workspace

Reply:
Oh okay, sorry I misunderstood. In my head you wanted to create a `.west` folder at the manifest path directory.

I updated the code, the structure look like this:
```
zephyrproject/                          
├── .west/                              
│   └── fish                            
│       └── [manifest_path_hash]/       
│           └── fish_boards_completion.cache
├── zephyr/                             
└── [...]                               
```

One small problem with that is that it add a call to `west topdir` but I tested it and it's still acceptable IMO